File: packages/ts/src/testing/fakes.ts
// Инъектируемые фейки для юнит-тестов агентов (Уровень 1 — без сети).
import { BillingExecutionDeniedError } from '../billing/errors'
import { hasRequiredAccess } from '../billing/client'
import type {
  BillingClient,
  BillingExecutionRequirement,
  BillingRuntimeState,
  BillingUsageEventInput,
} from '../billing/types'
import type { Claims, JwtVerifier } from '../auth/types'

/** Верификатор, возвращающий заданные claims без проверки подписи (только для тестов). */
export class FakeJwtVerifier implements JwtVerifier {
  constructor(private readonly claims: Claims) {}
  async verify(_token: string): Promise<Claims> {
    return this.claims
  }
}

export interface SentUsage {
  transactionId: string
  code: string
  properties: Record<string, unknown>
}

/** In-memory billing-клиент: отдаёт фикстуру runtime state, пишет usage в .sentUsage. */
export class InMemoryBillingClient implements BillingClient {
  readonly sentUsage: SentUsage[] = []
  private state: BillingRuntimeState

  constructor(options: { runtimeState: BillingRuntimeState }) {
    this.state = options.runtimeState
  }

  setRuntimeState(state: BillingRuntimeState): void {
    this.state = state
  }

  async getRuntimeStateByBillingOrgId(
    _billingOrgId: string,
  ): Promise<BillingRuntimeState> {
    return this.state
  }

  async assertExecutionAllowed(
    _billingOrgId: string,
    requirement?: BillingExecutionRequirement,
  ): Promise<BillingRuntimeState> {
    const s = this.state
    if (
      s.entitlementStatus !== 'active' ||
      s.remainingTotalTokens <= 0 ||
      !hasRequiredAccess(s, requirement)
    ) {
      throw new BillingExecutionDeniedError(s)
    }
    return s
  }

  async sendUsageEvent(event: BillingUsageEventInput): Promise<void> {
    this.sentUsage.push({
      transactionId: event.transactionId,
      code: event.code,
      properties: event.properties ?? {},
    })
  }
}

File: packages/python/src/ai37_agent_sdk/testing/fakes.py
from __future__ import annotations

from typing import Any

from ..auth.types import Claims
from ..billing.client import has_required_access
from ..billing.errors import BillingExecutionDeniedError
from ..billing.types import (
    BillingExecutionRequirement,
    BillingRuntimeState,
    BillingUsageEventInput,
)


class FakeJwtVerifier:
    """Возвращает заданные claims без проверки подписи (только для тестов)."""

    def __init__(self, claims: Claims) -> None:
        self._claims = claims

    def verify(self, token: str) -> Claims:  # noqa: ARG002 — токен игнорируется
        return self._claims


class InMemoryBillingClient:
    """In-memory billing-клиент: отдаёт фикстуру runtime state, пишет usage в ``sent_usage``."""

    def __init__(self, *, runtime_state: BillingRuntimeState) -> None:
        self._state = runtime_state
        self.sent_usage: list[dict[str, Any]] = []

    def set_runtime_state(self, state: BillingRuntimeState) -> None:
        self._state = state

    def get_runtime_state_by_billing_org_id(
        self, billing_org_id: str  # noqa: ARG002
    ) -> BillingRuntimeState:
        return self._state

    def assert_execution_allowed(
        self,
        billing_org_id: str,  # noqa: ARG002
        requirement: BillingExecutionRequirement | None = None,
    ) -> BillingRuntimeState:
        s = self._state
        if (
            s.entitlement_status != "active"
            or s.remaining_total_tokens <= 0
            or not has_required_access(s, requirement)
        ):
            raise BillingExecutionDeniedError(s, requirement)
        return s

    def send_usage_event(self, event: BillingUsageEventInput) -> None:
        self.sent_usage.append(
            {
                "transaction_id": event.transaction_id,
                "code": event.code,
                "properties": event.properties or {},
            }
        )
